### PR TITLE
Tamara solution 1- late but perhaps worth the wait?

### PR DESCRIPTION
**Pull request by tamarasaqer**
Shout out Raed for the idea.
---
name: solution review 
about: A template PR for code review with a checklist
---

### Files

- [x] The file name describes the function's behavior
- [x] There is a module docstring in the function file
- [x] The test file's name matches the function file name -
  `/tests/test_file_name.py`
- [x] There is a module docstring in the tests file

### Unit Tests

- [x] The test class has a helpful name in PascalCase
- [x] The test class has a docstring
- Every unit test has
  - [x] A helpful name
  - [x] A clear docstring
  - [x] Only one assertion
  - [x] There is no logic in the unit test
- [x] All tests pass
- [x] There are tests for defensive assertions
- [x] There are tests for boundary cases

### Function Docstring

- [x] The function's behavior is described
- The function's arguments are described:
  - [x] Type
  - [x] Purpose
  - [ ] Other assumptions (eg. if it's a number, what's the expected range?)
- The return value is described
  - [x] Type
  - [ ] Other assumptions are documented
- The defensive assertions are documented using `Raises:`
  - [ ] Each assumption about an argument is checked with an assertion
  - [x] Each assertion checks for _only one_ assumption about the argument
- [x] Include 3 or more (passing!) doctests

### The Function

- [x] The function's name describes it's behavior
- [x] The function's name matches the file name
- [x] The function has correct type annotations

### Do's

- [x] Variable names help to understand the strategy
- [ ] Any comments are clear and describe the strategy
- [x] Lines of code are spaced to help show different stages of the strategy

### Don'ts

- [x] The function's strategy _is not_ described in the documentation
- [ ] Comments explain the _strategy_, **not** the _implementation_
- [x] The function _does not_ have more comments than code
  - If it does, consider finding a new strategy or a simpler implementation

## Implementation

- [x] The code passes the formatting checks
- [x] The code passes all Ruff linting checks
- [x] The code has no (reasonable) Pylint errors
  - In code review, you can decide when fixing a Pylint error is helpful and
    when it's too restricting.
- [x] Variables are named with snake_case
- [x] Variable names are clear and helpful
- [x] The code follows the strategy as simply as possible
- [x] The implementation is as simple as possible given the strategy
- [x] The code includes defensive assertions
- [x] Defensive assertions include as little logic as possible

### CODE REVIEW

File: solutions/tests/test_is_odd_or_even.py [hunk 1]
```diff
@@ -0,0 +1,27 @@
+import unittest
```
Comment (AhmedKhalifa7) [inline solutions/tests/test_is_odd_or_even.py:1]: There is no module docstring in the tests file

File: solutions/tests/test_is_odd_or_even.py [hunk 2]
```diff
@@ -0,0 +1,27 @@
...
+from solutions.is_odd_or_even import is_odd_or_even
+
+
+class TestIsOddOrEven(unittest.TestCase):
+    """Unit tests for the is_odd_or_even function."""
+
+    def test_even_numbers(self):
+        """Test that even numbers return 'Even'."""
+        self.assertEqual(is_odd_or_even(4), "Even")
+        self.assertEqual(is_odd_or_even(0), "Even")
```
Comment (AhmedKhalifa7) [inline solutions/tests/test_is_odd_or_even.py:16]: Every unit test should have only one assertion

File: solutions/tests/test_is_odd_or_even.py [hunk 3]
```diff
@@ -0,0 +1,27 @@
...
+        self.assertEqual(is_odd_or_even(-2), "Even")
+
+    def test_odd_numbers(self):
+        """Test that odd numbers return 'Odd'."""
+        self.assertEqual(is_odd_or_even(7), "Odd")
+        self.assertEqual(is_odd_or_even(-3), "Odd")
```
Comment (AhmedKhalifa7) [inline solutions/tests/test_is_odd_or_even.py:22]: Each unit test should have only one assertion

File: solutions/is_odd_or_even.py [hunk 4]
```diff
@@ -0,0 +1,29 @@
+def is_odd_or_even(number):
+    """
+    Check if a number is odd or even.
+
+    Args:
+        number (int): The number to check.
+
+    Returns:
+        str: "Even" if the number is even, "Odd" if the number is odd.
+
```
Comment (AhmedKhalifa7) [inline solutions/is_odd_or_even.py:10]: If you can, add defensive assertions

File: solutions/tests/test_is_odd_or_even.py [hunk 5]
```diff
@@ -0,0 +1,27 @@
...
+        self.assertEqual(is_odd_or_even(1), "Odd")
+
```
Comment (AhmedKhalifa7) [inline solutions/tests/test_is_odd_or_even.py:24]: If you can, add tests for defensive assertions

File: solutions/is_odd_or_even.py [hunk 6]
```diff
@@ -0,0 +1,29 @@
+def is_odd_or_even(number):
+    """
```
Comment (AhmedKhalifa7) [inline solutions/is_odd_or_even.py:2]: There is no module docstring in the function file

File: solutions/is_odd_or_even.py [hunk 7]
```diff
@@ -0,0 +1,29 @@
+def is_odd_or_even(number):
```
Comment (AhmedKhalifa7) [inline solutions/is_odd_or_even.py:1]: The function has no type annotations

File: solutions/is_odd_or_even.py [hunk 8]
```diff
@@ -0,0 +1,29 @@
...
+        >>> is_odd_or_even(7)
+        'Odd'
+        >>> is_odd_or_even(0)
+        'Even'
+        >>> is_odd_or_even(-3)
+        'Odd'
+        >>> is_odd_or_even(-2)
+        'Even'
+    """
+    return "Even" if number % 2 == 0 else "Odd"
```
Comment (AhmedKhalifa7) [inline solutions/is_odd_or_even.py:23]: If you can, add defensive assertions

**Review by AhmedKhalifa7 — CHANGES_REQUESTED** [review]
Please address these changes. 
If you need any help just tell us.

File: solutions/tests/test_is_odd_or_even.py [hunk 9]
```diff
@@ -0,0 +1,27 @@
+import unittest
+import sys
+import os
+
+sys.path.insert(0, os.path.abspath(os.path.join(os.path.dirname(__file__), "../")))
```
Comment (RaedAlyan) [inline solutions/tests/test_is_odd_or_even.py:5]: why did you add this line?

**Comment by tamarasaqer** [comment]
> ## Kindly, provide a docstring at the start of the Python file that includes:
> 1- brief description of the module's purpose.
> 2- the author's name.
> 3- the date it was written.
> 
> This is a good practice to follow, as it helps maintain clear and informative code documentation.

Thank you, Mohammed! I agree. 

> ## Kindly, provide a docstring at the start of the Python file that includes:
> 1- brief description of the module's purpose.
> 2- the author's name.
> 3- the date it was written.
> 
> This is a good practice to follow, as it helps maintain clear and informative code documentation.

Thank you, Mohammad! I’ve updated the code to include the docstring with the required details. Appreciate the reminder!

**Comment by tamarasaqer** [comment]
> ## Kindly, provide a docstring at the start of the Python file that includes:
> 1- brief description of the module's purpose.
> 2- the author's name.
> 3- the date it was written.
> 
> This is a good practice to follow, as it helps maintain clear and informative code documentation.

Thank you, Mohammad! I’ve updated the code to include the docstring with the required details. Appreciate the reminder!

**Comment by tamarasaqer** [comment]
> Please address these changes.
> If you need any help just tell us.

Thank you, Ahmed! I’ve made the changes and updated the code as requested.

File: solutions/is_odd_or_even.py (hunk 6 above)
Comment (AhmedKhalifa7) [reply solutions/is_odd_or_even.py:14]: > There is no module docstring in the function file

Everything looks good now!

File: solutions/is_odd_or_even.py (hunk 4 above)
Comment (AhmedKhalifa7) [reply solutions/is_odd_or_even.py:22]: > If you can, add defensive assertions

Everything looks good now!

File: solutions/tests/test_is_odd_or_even.py (hunk 1 above)
Comment (AhmedKhalifa7) [reply solutions/tests/test_is_odd_or_even.py:11]: > There is no module docstring in the tests file

Everything looks good now!

File: solutions/tests/test_is_odd_or_even.py (hunk 5 above)
Comment (AhmedKhalifa7) [reply solutions/tests/test_is_odd_or_even.py:74]: > If you can, add tests for defensive assertions

Everything looks good

**Review by AhmedKhalifa7 — CHANGES_REQUESTED** [review]
A tiny change 🤏, other than this everything as it should be!

**Review by MohammadRAlSalloum — APPROVED** [review]
Great Work and all comments are resolved

**Review by AhmedKhalifa7 — APPROVED** [review]
:shipit: !!!!!!!!!!

File: solutions/is_odd_or_even.py [hunk 10]
```diff
@@ -0,0 +1,46 @@
...
+        >>> is_odd_or_even(7)
+        'Odd'
+        >>> is_odd_or_even(0)
+        'Even'
+        >>> is_odd_or_even(-3)
+        'Odd'
+        >>> is_odd_or_even(-2)
+        'Even'
+    """
+    if not isinstance(number, int):
```
[the remark this answers is not in the document]
Comment (AhmedKhalifa7) [reply solutions/is_odd_or_even.py]: Everything is good now

File: solutions/is_odd_or_even.py (hunk 7 above)
Comment (AhmedKhalifa7) [reply solutions/is_odd_or_even.py]: Everything is good now

File: solutions/tests/test_is_odd_or_even.py (hunk 3 above)
Comment (AhmedKhalifa7) [reply solutions/tests/test_is_odd_or_even.py]: Couldn't be better now
